qualtran/bloqs/arithmetic/controlled_addition.py

```diff
@@ -0,0 +1,240 @@
+        return self.a_dtype
+
+    @a_dtype.validator
+    def _a_dtype_validate(self, field, val):
+        if not isinstance(val, (QInt, QUInt, QMontgomeryUInt)):
+            raise ValueError("Only QInt, QUInt and QMontgomerUInt types are supported.")
+        if isinstance(val.num_qubits, sympy.Expr):
+            return
+        if val.bitsize > self.b_dtype.bitsize:
+            raise ValueError("a_dtype bitsize must be less than or equal to b_dtype bitsize")
+
+    @b_dtype.validator
+    def _b_dtype_validate(self, field, val):
+        if not isinstance(val, (QInt, QUInt, QMontgomeryUInt)):
+            raise ValueError("Only QInt, QUInt and QMontgomerUInt types are supported.")
+
+    @controlled.validator
+    def _controlled_validate(self, field, val):
+        if val not in (0, 1):
+            raise ValueError("controlled must be either 0 or 1")
+
+    @property
+    def signature(self):
+        return Signature(
+            [Register("ctrl", QBit()), Register("a", self.a_dtype), Register("b", self.b_dtype)]
+        )
+
+    def add_my_tensors(
+        self,
+        tn: 'qtn.TensorNetwork',
+        tag: Any,
+        *,
+        incoming: Dict[str, 'SoquetT'],
+        outgoing: Dict[str, 'SoquetT'],
+    ):
+        import quimb.tensor as qtn
+
+        if isinstance(self.a_dtype, QInt) or isinstance(self.b_dtype, QInt):
+            raise TypeError("Tensor contraction for addition is only supported for unsigned ints.")
+        N_a = 2**self.a_dtype.bitsize
+        N_b = 2**self.b_dtype.bitsize
+        inds = (
+            incoming['ctrl'],
+            incoming['a'],
+            incoming['b'],
+            outgoing['ctrl'],
+            outgoing['a'],
+            outgoing['b'],
+        )
+        unitary = np.zeros((2, N_a, N_b, 2, N_a, N_b), dtype=np.complex128)
+        for c, a, b in itertools.product(range(2), range(N_a), range(N_b)):
+            if c == self.controlled:
+                unitary[c, a, b, c, a, int(math.fmod(a + b, N_b))] = 1
+            else:
+                unitary[c, a, b, c, a, b] = 1
+
+        tn.add(qtn.Tensor(data=unitary, inds=inds, tags=[self.short_name(), tag]))
+
+    def decompose_bloq(self) -> 'CompositeBloq':
+        return decompose_from_cirq_style_method(self)
+
+    def on_classical_vals(self, **kwargs) -> Dict[str, 'ClassicalValT']:
+        a, b = kwargs['a'], kwargs['b']
+        unsigned = isinstance(self.a_dtype, (QUInt, QMontgomeryUInt))
+        b_bitsize = self.b_dtype.bitsize
+        N = 2**b_bitsize if unsigned else 2 ** (b_bitsize - 1)
+        ctrl = kwargs['ctrl']
+        if ctrl != self.controlled:
+            return {'ctrl': ctrl, 'a': a, 'b': b}
+        else:
+            return {'ctrl': ctrl, 'a': a, 'b': int(math.fmod(a + b, N))}
+
+    def short_name(self) -> str:
+        return "a+b"
+
+    def _circuit_diagram_info_(self, _) -> cirq.CircuitDiagramInfo:
+        if self.controlled is not None:
+            wire_symbols = ["In(ctrl)"]
```

Comment:
Double check this is the correct wire symbol for ctrl. Typically I read In as input, which I guess the ctrl value is but it's sort of a separate concept.

Reply:
I am pretty sure that there is a wire_symbol test:
    expected_wire_symbols = (
        ("In(ctrl)",) + ("In(x)",) * num_bits_a + ("In(y)/Out(x+y)",) * num_bits_b
    )
    assert cirq.circuit_diagram_info(gate).wire_symbols == expected_wire_symbols

Reply:
this method was removed per Matthew's comment